File: Trabajo_9/CPP/Actividades/Actividad_4.cpp
#include <iostream>
using namespace std;
//class interfaz - base
class Animal{
public:
    virtual void hacerSonido() = 0;
    virtual ostream& operator<<(ostream& os) const = 0;
};

ostream& operator<<(ostream& os, const Animal& animal) {
    return animal.operator<<(os);
}

class Perro : public Animal{
private:
    string nombre;
    int edad;

public:
    Perro(string n, int a){
        this -> nombre = n;
        this -> edad = a;
    }

    void hacerSonido() override{
        cout << "Miau" << endl;
    }

    ostream& operator<<(ostream& os) const override {
        os << "Nombre -> " << nombre << endl;
        os << "Edad -> " << edad << endl;
        return os;
    }

};

class Gato : public Animal{
private:
    string nombre;
    int edad;

public:
    Gato(string n, int a){
        this -> nombre = n;
        this -> edad = a;
    }

    void hacerSonido() override{
        cout << "Guau" << endl;
    }

    ostream& operator<<(ostream& os) const override {
        os << "Nombre -> " << nombre << endl;
        os << "Edad -> " << edad << endl;
        return os;
    }
};

class Vaca : public Animal{
private:
    string nombre;
    int edad;

public:
    Vaca(string n, int a){
        this -> nombre = n;
        this -> edad = a;
    }

    void hacerSonido() override{
        cout << "Invadamos Polonia" << endl;
    }

    ostream& operator<<(ostream& os) const override {
        os << "Nombre -> " << nombre << endl;
        os << "Edad -> " << edad << endl;
        return os;
    }
};

int main(){
    Animal* perrito = new Perro("Mateo", 5);
    Animal* gatito = new Gato("Steve", 9);
    Animal* vaquita = new Vaca("Dario", 2);


    cout << *perrito ;






    return 0;
}

File: Trabajo_9/JAVA/Actividades/Actividad_4.java
package Trabajo_9.Actividades;

class Animal {
    public void hacerSonido() 
    {
        System.out.println("Sonido genérico de animal");
    }
}

class Perro extends Animal {
    @Override
    public void hacerSonido() 
    {
        System.out.println("Guau!");
    }
}

class Gato extends Animal{
    @Override
    public void hacerSonido()
    {
        System.out.println("Miau");
    }

}

class Vaca extends Animal{
    @Override
    public void hacerSonido()
    {
        System.out.println("Muuu");
    }

}

public class Actividad_4 {
    public static void main(String[] args) {
    
        Animal[] animales = new Animal[3];

        animales[0] = new Perro();
        animales[1] = new Gato();
        animales[2] = new Vaca();

        for (Animal animal : animales) {
            animal.hacerSonido();
        }   
    }
}  